src/gpuarray_reduction.c

```diff
@@ -0,0 +1,548 @@
+	strb_appends(s, "\t\n");
+	
+	/**
+	 * Inner Loop Header Generation
+	 */
+	
+	for(i=ctx->numFreeIdx;i<ctx->numIdx;i++){
+		strb_appendf(s, "\tFOROVER(%d){ESCAPE(%d)\n", i, i);
+	}
+	
+	/**
+	 * Inner Loop Body Generation
+	 */
+	
+	appendIdxes (s, "\tT V = SRCINDEXER(", "i", 0, ctx->numIdx, "", ");\n");
+	strb_appends(s, "\t\n");
+	strb_appends(s, "\tif(V > maxV){\n");
+	strb_appends(s, "\t\tmaxV = V;\n");
+	appendIdxes (s, "\t\tmaxI = RDXINDEXER(", "i", ctx->numFreeIdx, ctx->numIdx, "", ");\n");
+	strb_appends(s, "\t}\n");
+	
+	/**
+	 * Inner Loop Trailer Generation
+	 */
+	
+	for(i=ctx->numFreeIdx;i<ctx->numIdx;i++){
+		strb_appends(s, "\t}\n");
+	}
+	strb_appends(s, "\t\n");
+	
+	/**
+	 * Inner Loop Epilogue Generation
+	 */
+	
+	strb_appends(s, "\t/**\n");
+	strb_appends(s, "\t * Destination writeback.\n");
+	strb_appends(s, "\t */\n");
+	strb_appends(s, "\t\n");
+	appendIdxes (s, "\tDSTMINDEXER(", "i", 0, ctx->numFreeIdx, "", ") = maxV;\n");
+	appendIdxes (s, "\tDSTAINDEXER(", "i", 0, ctx->numFreeIdx, "", ") = maxI;\n");
+}
+static void  appendLoopMacroUndefs  (strb*           s,
+                                     GEN_KERNEL_CTX* ctx){
+	strb_appends(s, "\t#undef FOROVER\n");
+	strb_appends(s, "\t#undef ESCAPE\n");
+	strb_appends(s, "\t#undef SRCINDEXER\n");
+	strb_appends(s, "\t#undef RDXINDEXER\n");
+	strb_appends(s, "\t#undef DSTMINDEXER\n");
+	strb_appends(s, "\t#undef DSTAINDEXER\n");
+}
+
+
+/**
+ * FIXME: Implement a working scheduler and invoker.
+ * 
+ * To schedule effectively the work across several dimensions of possibly
+ * ugly numbers, taking into account the limitations on thread & block
+ * scheduling, will require some library capable of providing a "factoring"
+ * of the tensor dimensions into "nice" prime numbers. Their product may
+ * be allowed to be larger than the original number, provided it remains
+ * within bounds.
+ */
+
+/**
+ * Compute a good thread block size / grid size for Nvidia.
+ */
+
+static void  scheduleMaxAndArgmax   (size_t*         blockSize,
+                                     size_t*         gridSize,
+                                     const GpuArray* src,
+                                     const int*      isReduced){
+	//int maxThreadPerBlock = 1024;
+	//int numFreeIdx = src->nd - getRdxIdx(src->nd, isReduced);
+	
+	/* Naive solution. Optimization is a tough problem. */
+	blockSize[0] = blockSize[1] = blockSize[2] = 1;
+	gridSize [0] = gridSize [1] = gridSize [2] = 1;
+}
+
+/**
+ * Invoke the kernel.
+ */
+
+static void  invokeMaxAndArgmax     (GpuKernel*      kernel,
+                                     const GpuArray* src,
+                                     const int*      isReduced){
+	size_t blockSize[3];
+	size_t gridSize[3];
+	
+	scheduleMaxAndArgmax(blockSize, gridSize, src, isReduced);
+	GpuKernel_call(kernel,
+	               getRdxIdx(src->nd, isReduced),
+	               blockSize,
+	               gridSize,
+	               0,
+	               NULL);
```

Comment:
You don't pass any arguments to the kernel anywhere.  I don't see how this can work.